[INT-3666] - Implementation of Metadastore for MongoDB

senthilaru opened the pull request:
[INT-3666] - Implementation of Metadastore for MongoDB

senthilaru committed:
MetadataStore for MongoDB using ConcurrentMetastore

senthilaru committed:
[INT-3666] - Implementation of Metadastore for MongoDB - added
standards/copyrights

senthilaru committed:
[INT-3666] - Testcases for Implementation of Metadastore for MongoDB

senthilaru committed:
[INT-3666] - Testcases MongoDB Metadatastore

senthilaru commented:
I have signed and agree to the terms of the SpringSource Individual
Contributor License Agreement.

artembilan commented:
@senthilaru , thank you for the contribution, anyway!

Let us take a look what you suggest!
But, first of all: does it say anything to you?
```
org.springframework.integration.mongodb.metadata.MongoDbMetadataStoreTests > testPutIfAbsent FAILED

java.lang.NullPointerException at MongoDbMetadataStoreTests.java:65

org.springframework.integration.mongodb.metadata.MongoDbMetadataStoreTests > testReplace FAILED

java.lang.AssertionError at MongoDbMetadataStoreTests.java:103

org.springframework.integration.mongodb.metadata.MongoDbMetadataStoreTests > testGetFromStore FAILED

java.lang.NullPointerException at MongoDbMetadataStoreTests.java:49

org.springframework.integration.mongodb.metadata.MongoDbMetadataStoreTests > testRemove FAILED

java.lang.NullPointerException at MongoDbMetadataStoreTests.java:80
```

senthilaru commented:
I m looking into the test case failure, but its working good in my local..!!

artembilan commented:
> but its working good in my local..!!

M-m-m. Do you really sure that you have MongoDB locally and it is run on the default `27017` port ?

Another concern: I'd prefer bypass that Domain object for the metadata entry.

senthilaru committed:
[INT-3666] - Applied null check and Testcases changes

senthilaru commented:
I have fixed the test case failure in the branch INT-3666, but Travis build is failing as its referring to master branch. 
I m working on removing the domain object and make use of BasicDBObject, but its not allowing dot(.) in the key( Key in this case will be /path/tofile/fileName.txt).

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,136 @@
+
+	private MongoTemplate template;
+	private final static String DEFAULT_COLLECTION_NAME = "metadatastore";
+	private String collectionName;
```
These fields must be `final`. 
I also see some code style vulnerabilities, but that can be fixed on merge.
BTW, @senthilaru , would be good to know your real name, like in passport :smile: .

senthilaru replied:
Thats my real name as per passport - Senthil Arumugam, Samiraj Panneer Selvam. I was tried to use that as SenthilArumugam SP, but its mentioned to give my full name :-)

artembilan replied:
Interesting. Can you say, please, what is fist name and what is last name?

senthilaru replied:
First Name: Senthil Arumugam
Last Name: Samiraj Panneer Selvam

artembilan replied:
Great! Thanks.
Doesn't Git allow you to do that like you shown?
```
Senthil Arumugam, Samiraj Panneerselvam
```
And the same for the `@author` tag.

Out of topic: `@since 4.2` :smile:

senthilaru replied:
Fixed

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,136 @@
+	 * 
+	 * @param template
+	 */
+	public MongoDbMetadataStore(MongoTemplate template) {
```
I think an additional `MongoDbFactory` ctor would be useful too.

senthilaru replied:
Fixed

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,136 @@
+	 * Configures the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
+	 * @param template
```
JavaDocs must be finished. Java 8 doesn't allow empty `@param`

senthilaru replied:
Fixed

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,136 @@
+		Datastore fileInfo = new Datastore(key, value);
+		if (!template.exists(query, collectionName)) {
+			template.save(fileInfo, collectionName);
+		}
```
No, the `save` for `put` is enough. See `RedisMetadataStore` as an example.
Any `put` to the `MetadataStore` should store requested value for the `key`. That's all.
Thanks to the MongoDB `save` operation we have:
```
* Update an existing document or insert a document depending on the parameter. If the document does not contain an '_id' field, then
* the method performs an insert with the specified fields in the document as well as an '_id' field with a unique objectid value. If
* the document contains an '_id' field, then the method performs an upsert querying the collection on the '_id' field: <ul> <li>If a
* document does not exist with the specified '_id' value, the method performs an insert with the specified fields in the document.</li>
* <li>If a document exists with the specified '_id' value, the method performs an update, replacing all field in the existing record
* with the fields from the document.</li> </ul>
```
From here I'd prefer to make that `key` as a `_id` field:
```
* If you object has an "Id' property, it will be set with the generated Id from MongoDB. If your Id property is a
* String then MongoDB ObjectId will be used to populate that string. 
```

And yes, I think the `BasicDBObject` is the best choice for this metadata domain model. With two field: `_id` and `value`

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,136 @@
+		if(result != null){
+			return result.value;	
+		}
+		return null;
```
Seems for me this would be enough:
````java
Query query = Query.query(Criteria.where("_id").is(key));
query.fields().include("value");
return (String) template.findOne(query, Map.class, collectionName).get("value");
````

senthilaru replied:
Fixed.

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,136 @@
+		Assert.notNull(key, "'key' must not be null.");
+		Query query = new Query(Criteria.where("key").is(key));
+		Datastore result = template.findAndRemove(query, Datastore.class,collectionName);
+		return result==null?null:result.value;
```
The same is here for that `include("value")` and `Map` result

senthilaru replied:
Fixed.

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,136 @@
+		} else {
+			template.save(new Datastore(key, value), collectionName);
+		}
+		return result;
```
`insert` exists for such an operation.

artembilan replied:
From other side this logic doesn't follow with `putIfAbsent` contract:
```
 * If the specified key is not already associated
 * with a value, associate it with the given value.
 * This is equivalent to
 *  <pre> {@code
 * if (!map.containsKey(key))
 *   return map.put(key, value);
 * else
 *   return map.get(key);
 * }</pre>
 *
 * except that the action is performed atomically.
```

Unfortunately I don't see any similar MongoDB atomic operation. So we might end up with something like:
````java
try {
     template.insert(entry, collectionName);
}
catch (MongoDataIntegrityViolationException e)  {
    return get(key);
}
return null;
````
CC @christophstrobl

senthilaru replied:
Fixed.

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,136 @@
+		Assert.notNull(oldValue, "'oldValue' must not be null.");
+		Assert.notNull(newValue, "'newValue' must not be null.");
+		
+		Query query = new Query(Criteria.where("key").is(key));
```
You have missed `oldValue` in the criteria.
See the `replace` contract in the `ConcurrentMap.replace`

senthilaru replied:
Fixed

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,136 @@
+		return result.isUpdateOfExisting();
+	}
+
+	private class Datastore {
```
As we talked this is an overhead. The simple `HashMap` can be used to store and `Map.class` to read.

senthilaru replied:
Fixed.

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/package-info.java:
```diff
@@ -0,0 +1,8 @@
+ * 
+ */
+/**
+ * @author SenthilArumugam SP
```
`package-info.java` must not contains an author name, but JavaDocs is required.
As well as new line in the end of each file.

senthilaru replied:
Fixed

artembilan commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStoreTests.java:
```diff
@@ -0,0 +1,104 @@
+	
+	@Test
+	@MongoDbAvailable
+	@Before
```
If this is `@Before` why it is `@Test`, too?
Bug ?

senthilaru replied:
Fixed

senthilaru committed:
[INT-3666] - Author changed, param desc added, replace contract fixed,
put() without any check.

senthilaru commented:
Still, working on the major fixes.,

senthilaru committed:
[INT-3666] - Removed DataStore inner class overhead and utilized Map.
Key as _Id (default indexing), putIfAbsent - changed to atomic mongo
operation,

senthilaru committed:
[INT-3666] - Removed diamond operator in favour of Java1.6

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+
+	protected final Log logger = LogFactory.getLog(getClass());
+	private final MongoTemplate template;
+	private final String DEFAULT_COLLECTION_NAME = "metadatastore";
```
Hi @senthilaru ! Thank you for the update!

Can't we have this default collection name in the camel case?

Nit-pick here for the code style: blank lines between class fields. Constants - first and `static` as well. The `logger` can be static too.

senthilaru replied:
I was getting warning in the checkstyle review - "'static' modifier out of order with the JLS suggestions." So, i removed static., what do you suggest, shall we add it?

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+	 * 
+	 * Configures the MongoDbMetadataStore by provided {@link MongoTemplate} and
+	 * default collection name - {@link #DEFAULT_COLLECTION_NAME}.
+	 * 
```
No blank lines in the method JavaDocs. Minor, of course, and can be polished on merge. If you aren't going to contribute anymore :smile:

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+	 */
+	public MongoDbMetadataStore(MongoTemplate template, String collectionName) {
+		this.template = template;
+		this.collectionName = (StringUtils.hasText(collectionName)) ? collectionName
```
I'd say that must be vise versa: ctor without `collectionName` should use this one with `DEFAULT_COLLECTION_NAME`. And here must be
```
Assert.hasText(collectionName);
Assert.notNull(template);
```
with appropriate error message to warn end-user against the wrong actions.

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+	}
+
+	/**
+	 * Configures the MongoDbMetadataStore by provided {@link MongoTemplate} and
```
The method JavaDocs must be `imperative`, In other words `Configure` not  `Configures`

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+		template = new MongoTemplate(factory);
+		this.collectionName = (StringUtils.hasText(collectionName)) ? collectionName
+				: DEFAULT_COLLECTION_NAME;
+	}
```
This ctor has to invoke the other one with similar logic as I described above. DRY - in other words.

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+	 * Method put.
+	 * @param key String
+	 * @param value String
+	 * @see org.springframework.integration.metadata.MetadataStore#put(String, String)
```
Please, don't forget that Java 8 is strict for the JavaDocs. And describe, please, here which MongoDB operation performs this method.

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+	public void put(String key, String value) {
+		Assert.notNull(key, "'key' must not be null.");
+		Assert.notNull(value, "'value' must not be null.");
+		Map<String,String> fileInfo = new HashMap();
```
Nit-pick for the variable name. It is really not `file`, so just `entry` would be good.
The `new HashMap()` is bad code too. Must be with generics

artembilan commented:
General comment: please, rebase your branch to the last `master`: it is now out of date.

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+	 * Method get.
+	 * @param key String
+	 * @return String
+	 * @see org.springframework.integration.metadata.MetadataStore#get(String)
```
The same concern here for the JavaDocs

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+	@Override
+	public String get(String key) {
+		Assert.notNull(key, "'key' must not be null.");
+		Query query = new Query(Criteria.where(ID_FIELD).is(key));
```
Please, add here 
```
query.fields().exclude(ID_FIELD);
```
to increase the network throughput a bit.

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+		Query query = new Query(Criteria.where(ID_FIELD).is(key));
+		@SuppressWarnings("unchecked")
+		Map<String,String> result = template.findOne(query, Map.class, collectionName);
+		return result==null?null:result.get(VALUE);
```
The bad code format. Whitespaces must be there around operators. As well as in the generics definition after `,`

senthilaru replied:
It will be great, if we have common formatter or checkstyle.!!

artembilan replied:
It isn't so easy to provide such a tool, hence we just follow with code style rules: https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-Code-Style.
Some of them you can import to your IDE: 
https://github.com/spring-projects/spring-framework/tree/master/src/idea
https://github.com/spring-projects/spring-framework/tree/master/src/eclipse
But of course, there is no such an option to format your whitespaces vulnerabilities :smile:

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+	@Override
+	public String remove(String key) {
+		Assert.notNull(key, "'key' must not be null.");
+		Query query = new Query(Criteria.where(ID_FIELD).is(key));
```
The same `query.fields().exclude(ID_FIELD);` is here too.

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+		Query query = new Query(Criteria.where(ID_FIELD).is(key));
+		@SuppressWarnings("unchecked")
+		Map<String,String> result = template.findAndRemove(query, Map.class,collectionName);
+		return result==null?null:result.get(VALUE);
```
The same code format issue is here too

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+			template.insert(fileInfo, collectionName);
+		} catch (DuplicateKeyException e) {
+			if(logger.isDebugEnabled()){
+				logger.debug("DUPLICATE KEY VIOLOATION",e);
```
Typo. However I don't think that we should say this `debug` at all, because it is normal behaviour for `putIfAbsent`  contract.
From other side we talked with @christophstrobl about this operation and found that it can be achieved only with the `raw MongoDB script execution` to be sure in the atomicity.
This feature is available since the last Spring Data Fowler and SD-Mongo-1.7.0.
Spring Integration is already based on the last SD, but you should rebase your branch to the last `master`.

How script looks we'll show you later: after fixing for all other concerns to this PR.

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+		Query query = new Query(Criteria.where(ID_FIELD).is(key).and(VALUE).is(oldValue));
+		Update update = new Update();
+		update.set(VALUE, newValue);
+		WriteResult result = template.updateFirst(query, update, Map.class,collectionName);
```
No reason in the `entityClass` here. We are interested in the `isUpdateOfExisting()` only.

senthilaru replied:
Fixed

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+		Assert.notNull(oldValue, "'oldValue' must not be null.");
+		Assert.notNull(newValue, "'newValue' must not be null.");
+		
+		Query query = new Query(Criteria.where(ID_FIELD).is(key).and(VALUE).is(oldValue));
```
I guess we should `ensureIndex` for this search.
For this purpose this class should implement `InitializingBean` and does this in the `afterPropertiesSet()`:
```
IndexOperations indexOperations = this.mongoTemplate.indexOps(this.collectionName);
indexOperations.ensureIndex(new Index(ID_FIELD, Sort.Direction.ASC)
              .on(VALUE, Sort.Direction.ASC));
```

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/package-info.java:
```diff
@@ -0,0 +1,4 @@
+/**
+ * Contains mongodb metadata store related classes
+ */
+package org.springframework.integration.mongodb.metadata;
```
New line in the end of each affected file.

senthilaru replied:
Fixed

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2002-2015 the original author or authors.
```
No reason of the `2002` for new classes. And blank line before Copyright.

senthilaru replied:
Fixed

artembilan commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStoreTests.java:
```diff
@@ -0,0 +1,136 @@
+
+/**
+ * @author Senthil Arumugam, Samiraj Panneer Selvam
+ *
```
`@since 4.2`

senthilaru replied:
Fixed.

artembilan commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStoreTests.java:
```diff
@@ -0,0 +1,136 @@
+	private MongoDbMetadataStore store = null;
+	
+	@MongoDbAvailable
+	@Before
```
No need for the `@MongoDbAvailable` on the `@Before`. 
From other side I don't see any reason in this  `@Before` method at all, since the logic of your test methods in many places are based on the fresh `MongoDbMetadataStore`

senthilaru replied:
We can remove @MongoDbAvailable, but we need @Before, because we are cleaning-up by invoking configure method before each test.

artembilan commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStoreTests.java:
```diff
@@ -0,0 +1,136 @@
+	
+	private final static String DEFAULT_COLLECTION_NAME = "metadatastore";
+	private final String file1 = "/remotepath/filesTodownload/file-1.txt";
+	private final String file1Id = "12345";
```
Since these look like constants they should be `static`.
What is the reason of that `DEFAULT_COLLECTION_NAME` ?

artembilan replied:
Ah! I see. for the `cleanupCollections`. OK. nevermind.

senthilaru replied:
Ignored.

artembilan commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStoreTests.java:
```diff
@@ -0,0 +1,136 @@
+	@MongoDbAvailable
+	@Test
+	public void testConfigureCustomCollection() throws Exception  {
+		String collectioName = "TestMetaStore";
```
Typo: collectio**n**Name

senthilaru replied:
you are very good in reviewing very closely., :+1:   I didn't notice it .,
Fixed

artembilan replied:
:smile: To be honest, an IDEA type plugin

artembilan commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStoreTests.java:
```diff
@@ -0,0 +1,136 @@
+	@MongoDbAvailable
+	@Test
+	public void testConfigureFactorCustomCollection() throws Exception  {
+		String collectioName = "TestMetaStore";
```
DITO

senthilaru replied:
Fixed.

artembilan commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStoreTests.java:
```diff
@@ -0,0 +1,136 @@
+	@Before
+	public void configure() throws Exception  {
+		MongoDbFactory mongoDbFactory = this.prepareMongoFactory(DEFAULT_COLLECTION_NAME);
+		MongoTemplate template = new MongoTemplate(mongoDbFactory);
```
Since we have already `MongoDbFactory` ctor this extra `MongoTemplate` variable is redundant

senthilaru replied:
Fixed.

artembilan commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStoreTests.java:
```diff
@@ -0,0 +1,136 @@
+	public void testReplace() throws Exception{
+		boolean removedValue = store.replace(file1, file1Id, "4567");
+		assertFalse(removedValue);
+		
```
I think here must be ` store.get(file1)` to be sure that the `replace` hasn't affected the collection.

senthilaru replied:
Fixed.

artembilan commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStoreTests.java:
```diff
@@ -0,0 +1,136 @@
+		assertNull(fileID);
+		
+		removedValue = store.replace(file1, file1Id, "4567");
+		assertTrue(removedValue);
```
And here `get` one more time to be sure that `replace` has changed the value in the document

senthilaru replied:
Fixed.

artembilan commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStoreTests.java:
```diff
@@ -0,0 +1,136 @@
+		fileID = store.putIfAbsent(file1, file1Id);
+		assertNull(fileID);
+		
+		fileID = store.putIfAbsent(file1, file1Id);
```
`putIfAbsent` for something different and `get` afterwards to be sure that `putIfAbsent` hasn't changed anything in the collection.

senthilaru replied:
Fixed.

artembilan commented on spring-integration-mongodb/src/test/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStoreTests.java:
```diff
@@ -0,0 +1,136 @@
+		
+		fileID = store.remove(file1);
+		assertNotNull(fileID);
+		assertEquals(file1Id, fileID);
```
And `get` in the end to check that there is nothing in the collection on he matter.

senthilaru replied:
Fixed.

artembilan commented:
@senthilaru , sorry for such a hypercritical review, but it is a way how we do in the Open Source.
Hope you'll study something new with us and will be more sure in the future contributions.

Thank you one more time for the PR !

artembilan commented:
@senthilaru , no reason to notify me about each fix. You can do that on the background and comment only if you don't agree with me or if you have other thoughts or concerns.

artembilan commented:
BTW, please, notify us just with comment `Pushed` when you add more commits to the PR.

senthilaru committed:
[INT-3666] - Added more test cases, validation in the ctr args., fixed
coding standards.

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,201 @@
+
+	protected final Log logger = LogFactory.getLog(getClass());
+	private final MongoTemplate template;
+	private final static String DEFAULT_COLLECTION_NAME = "metadataStore";
```
`private static final`, as for all other constants here: `ID_FIELD`, `VALUE`.
And I don't see whitespaces between fields.
And `static` field first in the class

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,201 @@
+		Assert.notNull(factory, "'factory' must not be null.");
+		Assert.hasText(collectionName, "'collectionName' must not be empty.");
+		template = new MongoTemplate(factory);
+		this.collectionName = collectionName;
```
`MongoTemplate` takes care about `factory` assert. Hence they are redundant here and this ctor must be like:
```
this(new MongoTemplate(factory), collectionName);
```
That's all!

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,201 @@
+		Assert.notNull(newValue, "'newValue' must not be null.");
+		Query query = new Query(Criteria.where(ID_FIELD).is(key).and(VALUE).is(oldValue));
+		Update update = new Update();
+		update.set(VALUE, newValue);
```
Can be simplified with `update` static factory method from the `Update`

artembilan commented:
@senthilaru , are you going to address other concerns, e.g. JavaDocs vulnerabilities ?
In addition, since it is a new feature to the Framework, we need Reference Manual update:
`whats-new.xml`, `mongodb.xml` and `meta-data-store.xml` with cross-links.
Would you mind doing that as well?

Thanks

senthilaru commented:
I m working on other things as well. I really appreciate your patience. I m learning a lot out of this..!!

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,201 @@
+		Map<String, String> entry = new HashMap<String, String>();
+		entry.put(ID_FIELD, key);
+		entry.put(VALUE, value);
+		template.save(entry, collectionName);
```
Looking at this closer I'd suggest this variant:
```
template.execute(collectionName, new CollectionCallback<Object>() {

		@Override
		public Object doInCollection(DBCollection collection) throws MongoException, DataAccessException {
			return collection.save(new BasicDBObject(entry));
		}

});
```
1. We bypass any Spring Data MongoDB stuff like events, conversions
2. We get rid of an additional `_class` document field

It isn't applied to the `putIfAbsent`, because we ill use there `script`.

BTW, please, rebase this branch to the latest master finally.

senthilaru committed:
[INT-3666] - Removed spring data mongo save

senthilaru committed:
[INT-3666] - Removed unnecessary docs.

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,198 @@
+	private static final String DEFAULT_COLLECTION_NAME = "metadataStore";
+	private final String collectionName;
+	private static final String ID_FIELD = "_id";
+	private static final String VALUE = "value";
```
Please, order field as I recommended: static first.
And add blank lines between them finally!

In addition, please, ensure that your IDE removes trailing whitespaces on save

artembilan commented on spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:
```diff
@@ -0,0 +1,198 @@
+	
+	/**
+	 * @param key String
+	 * @param value String
```
Please, add the comprehensive description for params.
And I recommend to add method description to address which MongoDB operation is performed

artembilan commented:
And, please, rebase your branch to the latest `master` finally. We can't provide the fix for the `putIfAbsent` without the new SD-Mongo version

senthilaru commented:
Rebase completed.

senthilaru committed:
[INT-3666] - Added Java docs Method and param description, code
formatting removed trail space, final added wherever possible.

artembilan commented:
Hi @senthilaru !

Thanks for the update! Sorry, if I wasn't clear, but all those new `final` modifiers are redundant.

Let me know if you are going to address documentation on the matter and want to take a look to the `script` solution for the `putIfAbsent` on your own. Otherwise we'll go ahead and do that for you.

senthilaru committed:
[INT-3666] - removed redundant finals

artembilan commented:
@senthilaru , I see some progress here.
Would you mind letting us know about other concerns? Are you going to address them? 
Or do you leave them for us?

Thanks

senthilaru commented:
@artembilan  - I tried different ways to achieve the atomic putIfAbsent, using mongodb data APIs, nothing worked out. Do you think any solution for it? Let me try this week, or else I will give it you guys :-)

artembilan commented:
Hi @senthilaru !

Thank you for looking for the solution!

As I said above we can do that only using `script` and new `ScriptOperations` since SD-Mongo-1.7.0.RC1.
Please, upgrade that one in the `build.gradle` and try to figure out how to be from the http://docs.spring.io/spring-data/data-mongo/docs/1.7.0.RC1/reference/html/#mongo.server-side-scripts, http://docs.mongodb.org/manual/reference/command/eval/
Let me know how it is for you. And answer, please, to the question regarding docs. Are you going to address them, too ?

senthilaru commented:
Pls. address the documentation issue or new xml files. I will do it in the next contribution., pls.
I will try with script & ScriptOperations in the RC version

artembilan commented:
Hi @senthilaru !

Hope you are well.

Please, don't do docs changes yet. We are working now under the task to covert our documentations to the Asciidoctor: #1407 . It would be a bit inconvenient to reconvert them one more time with your new additions regarding this MongoDB Module changes :smile: 

Thanks

artembilan commented:
@senthilaru , you have to rebase your branch if you are going to do docs changes.

We don't have that XML hell any more! But... switched to another: Asciidoctor :smile: 

Thanks

senthilaru committed:
[INT-3666] - Upgraded the SDMongo version to 1.7.0.RC1, atomic
putIfAbsent implemented with new API.

senthilaru commented:
Pls. review the changes. I noted that in the SD Mongo API, you should have handled the single quote introduction by JS. I have done it manually, let me know if can improve further, esp. register the JS once and called multiple times.

senthilaru committed:
[INT-3666] - Optimised the JS method call using NamedMongoScript

artembilan commented:
@senthilaru , thank you for the progress. 
But rebase, please, your branch to the `master`.
It's really very critical for such a change: we should be sure that it works with all new dependencies.

artembilan commented on build.gradle:
```diff
@@ -120,7 +120,7 @@ subprojects { subproject ->
 		smackVersion = '4.0.6'
 		springAmqpVersion = project.hasProperty('springAmqpVersion') ? project.springAmqpVersion : '1.5.0.BUILD-SNAPSHOT'
-		springDataMongoVersion = '1.7.0.M1'
+		springDataMongoVersion = '1.7.0.RC1'
```
Great! 
Thank you!
Now I can pull it locally and play.
FYI the Fowler RELEASE is here already: https://spring.io/blog/2015/03/23/spring-data-release-train-fowler-goes-ga
Hence we can upgrade just to `1.7.0.RELEASE`

artembilan commented:
@senthilaru , let me know if you are going to address docs.
And don't forget, please, notify us that you have done some push to the PR.

I'll back to you after some polishing work on this.

Thank you very much!

senthilaru committed:
[INT-3666] - upgraded SD-Mongo to release version

senthilaru commented:
Sorry Artembilian, I m not going to work on the docs. for now. pls. take care of it., as you are someone close to its convention/format changes.

artembilan commented:
No problem, @senthilaru !

I'll take care about docs and polishing to your work on Monday.

senthilaru commented:
Thank you so much..!!

artembilan commented:
The polishing for the review: https://github.com/artembilan/spring-integration/commit/04ab7d7c702c97078f38e8a89c3486ed0e64a22a

CC: @garyrussell

senthilaru commented:
any idea, when will it be merged and released?

garyrussell commented:
Thanks @senthilaru for your contribution!

We are anticipating a 4.2 milestone release in about a month.

Since there is no schema change we can consider back-porting this to 4.1.

Unfortunately, we just released 4.1.3 so it will be a few weeks before 4.1.4 will be released.

Due to unforeseen circumstances, I won't be able to review/merge until Wednesday; sorry about that.

garyrussell commented:
LGTM; aside from the one comment about `afterPropertiesSet()` in @artembilan 's polish.

Docs:

`s/All this operations are _atomic_ by the MongoDB guarantees./All these operations are _atomic_ via MongoDB guarantees./`

`s/For this purpose the putIfAbsent operation is implemented as as _stored_/...as a _stored_/`

Also, with what's new - just "a mongodb metadatastore is now available" and point to `<<mongodb-metadata-store>>` which already describes how it can be used. No point in partial reproduction here.

artembilan commented:
@garyrussell , the result after review: https://github.com/artembilan/spring-integration/commit/cfb007e2f7678df00227600136f3597ffb62bde1

Let me know and I'll rebase and merge that.

Thank you both!

artembilan commented:
And yes, with backport to 4.1.x (without Docs, of course)

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/d8f175edacf918d556e97093a9718c78536df19b

Thank you very much for the contribution one more time and we are looking forward for more.
Regarding backport, see my comment in JIRA
